R/tm_rmarkdown.R

```diff
@@ -0,0 +1,267 @@
+#'           },
+#'           n_rows_value = input$n_rows
+#'         )
+#'       })
+#'     })
+#'   }
+#' )
+#'
+#' app <- init(
+#'   data = data,
+#'   modules = modules(
+#'     tm_rmarkdown(
+#'       label = "RMarkdown Module",
+#'       rmd_file = system.file(file.path("sample_files", "test.Rmd"), package = "teal.modules.general"),
+#'       allow_download = FALSE,
+#'       extra_transform = list(nrow_transform)
+#'     )
+#'   )
+#' ) |> shiny::runApp()
+#' @export
+#'
+tm_rmarkdown <- function(label = "RMarkdown Module",
+                         rmd_file,
+                         datanames = "all",
+                         allow_download = TRUE,
+                         pre_output = NULL,
+                         post_output = NULL,
+                         transformators = list(),
+                         extra_transform = list()) {
+  message("Initializing tm_rmarkdown")
+
+  # Start of assertions
+
+  checkmate::assert_string(label)
+  checkmate::assert_file(rmd_file, access = "r")
+  checkmate::assert_flag(allow_download)
+
+  checkmate::assert_multi_class(pre_output, c("shiny.tag", "shiny.tag.list", "html"), null.ok = TRUE)
+  checkmate::assert_multi_class(post_output, c("shiny.tag", "shiny.tag.list", "html"), null.ok = TRUE)
+
+  # End of assertions
+
+  # Make UI args
+  args <- as.list(environment())
+
+  ans <- module(
+    label = label,
+    server = srv_rmarkdown,
+    server_args = list(rmd_file = rmd_file, allow_download = allow_download, extra_transform = extra_transform),
+    ui = ui_rmarkdown,
+    ui_args = args,
+    transformators = transformators,
+    datanames = datanames
+  )
+  # attr(ans, "teal_bookmarkable") <- TRUE
+  ans
+}
+
+# UI function for the rmarkdown module
+ui_rmarkdown <- function(id, rmd_file, allow_download, extra_transform, ...) {
+  args <- list(...)
+  ns <- NS(id)
+
+  teal.widgets::standard_layout(
+    output = teal.widgets::white_small_well(
+      tags$div(
+        tags$h4(
+          "Rendered report from: ",
+          tags$code(basename(rmd_file))
+        ),
+        if (allow_download) {
+          downloadButton(
+            ns("download_rmd"),
+            sprintf("Download '%s'", basename(rmd_file)),
+            class = "btn-primary btn-sm"
+          )
+        },
+        ui_transform_teal_data(ns("extra_transform"), transformators = extra_transform)
+      ),
+      tags$hr(),
+      uiOutput(ns("rmd_output"))
+    ),
+    encoding = NULL,
+    pre_output = args$pre_output,
+    post_output = args$post_output
+  )
+}
+
+# Server function for the rmarkdown module
+srv_rmarkdown <- function(id, data, rmd_file, allow_download, extra_transform) {
+  checkmate::assert_class(data, "reactive")
+  checkmate::assert_class(isolate(data()), "teal_data")
+  moduleServer(id, function(input, output, session) {
+    if (allow_download) {
+      output$download_rmd <- downloadHandler(
+        filename = function() basename(rmd_file),
```

Comment:
shouldn't downloadable rmarkdown contain also "reproducible" qenv code? IMO `tm_rmarkdown(rmd_file)` is just a template and returning a template doesn't make sense if it lacks data to actually execute templete-file.

Reply:
@llrs-roche parsing and adding the section in `.R` or in `.Rmd` would have similar challenges?

From my initial findings `knitr::purl()` is promising, but it does have some minor limitations with chunk options (it doesn't play well when evaluating params: `{r, eval = !is.null(params$CO2)}`)